### PR TITLE
CCS 163279: Added Shoe Size Templates #79

### PR DESCRIPTION
**Pull request by SakshamAccolite**
Added Shoe Size Templates

### CODE REVIEW

File: CloudFormationWindows/lansa-development-win.cfn.template [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        }
+    },
+   "Metadata" : {
+     "AWS::CloudFormation::Interface" : {
+       "ParameterGroups" : [
+         {
+           "Label" : { "default" : "Essential Parameters" },
+           "Parameters" : [ 
+               "01LansaMSI", 
+               "02LansaMSIBitness",
```
Comment (robe070) [inline CloudFormationWindows/lansa-development-win.cfn.template]: MSIBitness is meant to be hidden

**Review by robe070 — CHANGES_REQUESTED** [review]
Parameters must be removed from the Shoe Size templates, if they are hidden and do not change the default value. Its important that the Master template default values are used where possible so as to minimise changes in the future. e.g. `LANSAMSIBitness` & `22TriggerCakeUpdate` parameters must be removed. There are many others. Obviously parameters like `13DBInstanceClass` must be kept in order to change the default value.

I have checked the Development template thoroughly. I notice the default values for parameters has been correctly changed. I suggest this is the template you modify as described above.

LANSAVersion parameter will need to be added once the Master template has the parameter.

VPC parameters will need to be prompted for too.

I have not checked any of the other templates.

I suggest you make the alterations to just 1 template and submit it for my review before altering the other templates so you do not replicate incorrect coding. So it would be good to do these changes as early as possible today because I only have a 2.5 hour overlap with your normal work day. I will check it as soon as I see the notification. Posting in Slack as well would be good as I get those notifications when I'm away from my desk. I notice you are not on Slack, so ask Varun or another of your team to post for you.

**Comment by SakshamAccolite** [comment]
Hi @robe070 ,
I've made the changes suggested by you, Kindly review.

**Comment by robe070** [comment]
Remove these too. Unless you think they should be kept:
`02LansaMSIBitness`
`10LansaGitRepoBranch`
`17UserScriptHook`

**Comment by robe070** [comment]
...and at least start the execution of the template to prove that it is syntactically correct. Then delete the stack immediately.

**Comment by SakshamAccolite** [comment]
Yes @robe070 
02LansaMSIBitness can be removed( i missed), but 10LansaGitRepoBranch and 17UserScriptHook are being used in "conditions" and "resources", that's why i didn't remove. Orelse i'll have to use their hardcoded value on line 219, 229 and 235

**Comment by robe070** [comment]
`DoesNotHaveUserScript` & `HasUserScript` can be removed. They are not used.

**Comment by robe070** [comment]
As for the git branch, you will need to hard code it. Forget all the complicated options for determining the nested template location. Your task is just to put a veneer on the front of whatever version you are using.

I've already put a working version in `support/scalable` which is the production location. I've tested it. Obviously, it doesn't have the new vpc parameters. So that will need to be updated once they are available. But for now, your dev template will work with `support/scalable`. So thats your hard coded value. But, you can't upload your templates to that folder.

So, for the interim, and to be changed before you check it in, hard code its default to debug/paas. Keep the Git Branch parameter so its easy for you/[a developer] to change it. And I notice you are not supporting the hard coded 'source' value. Please talk to Varun about that.

So, as we need to keep the parameter, just leave the template look up as it is. Don't like it, but thats what we need to do.

**Comment by SakshamAccolite** [comment]
Hi @robe070 
If I enter the inputs given for "development" shoe size, and try to create stack(Using master template), i get below error:

RDS does not support creating a DB instance with the following combination: DBInstanceClass=db.t3.micro, Engine=sqlserver-web, EngineVersion=14.00.3281.6.v1, LicenseModel=license-included. For supported combinations of instance class and database engine version, see the documentation. (Service: AmazonRDS; Status Code: 400; Error Code: InvalidParameterCombination; Request ID: 236aa46c-b8de-4f14-8238-5043021f8c59; Proxy: null)

I guess this is not a valid resource combination.

**Comment by robe070** [comment]
Use db.t3.small

**Comment by SakshamAccolite** [comment]
Hi @robe070 ,
I've made the changes. Kindly review.

**Comment by robe070** [comment]
I see why you have done the changes the way you have given my comment above. Its confusing. Please put back in all the parameters for which you needed to change the default value:
```
                    "08RemoteAccessLocation": { "Ref" : "08RemoteAccessLocation" },
                    "10LansaGitRepoBranch": { "Ref" : "10LansaGitRepoBranch" },
                    "11WebserverOSVersion": { "Ref" : "11WebserverOSVersion" },
                    "11WebServerInstanceTyp": { "Ref" : "11WebServerInstanceTyp" },
                    "12WebServerMaxConnec": { "Ref" : "12WebServerMaxConnec" },
                    "13DBInstanceClass": { "Ref" : "13DBInstanceClass" },
                    "14DBName": { "Ref" : "14DBName" },
                    "15DBEngine": { "Ref" : "15DBEngine" },
                    "17UserScriptHook": { "Ref" : "17UserScriptHook" },
                    "18WebServerCapacity": { "Ref" : "18WebServerCapacity" },
                    "19DBAllocatedStorage": { "Ref" : "19DBAllocatedStorage" },
                    "20DBIops": { "Ref" : "20DBIops" },
```

And leave gitbranch as it was.

Managing each developer's templates separately and also setting it to be the correct value in production is not working well. Varun and I will need to discuss.

**Comment by SakshamAccolite** [comment]
Yes @robe070 I misinterpreted the above comment, sorry for that. I've added the required parameters back:)

**Comment by robe070** [comment]
Looks good. Have you created stacks with all the templates? Where did you upload the templates to?

**Comment by SakshamAccolite** [comment]
Yes @robe070 I created the stacks, and deleted immediately. I uploaded the templates here:
image

**Comment by robe070** [comment]
It looks like you didn't upload to your own S3 bucket. You've just let template creation create a temporary bucket for you. I'll do the same.

**Comment by robe070** [comment]
...and you'll need to incorporate changes from the Master template like the OS Version now has 3 values and the LANSA Version, and potentially other parameters. I think its best to leave those until they are complete. Just do it all in one go.

**Comment by robe070** [comment]
The order of the parameters and naming of groups of parameters has completely changed. This will take me a long time to sort through. I'll need to do it tomorrow.

**Comment by robe070** [comment]
Actually it looks quite easy. The original order is whats needed with original group naming (See Master template). Just do Development and I'll review what that looks like.

**Comment by robe070** [comment]
...and I don't mean that any parameters need to be added back. Its just the grouping and the order of the groups.

**Comment by robe070** [comment]
I've been just reading the code, and mistinterpreting it. Once I've run the template and in less of a rush on a Monday now, I see that my comments above are incorrect.

We need to go back to this commit : 2fb1836. That was mostly correct.
And fix up git branch to what it is in the latest commit.
And let me review that.

My profuse apologies.

**Comment by SakshamAccolite** [comment]
No worries @robe070. I think now it's what you expected. Just want to point out that 10LansaGitRepoBranch will be present in "Other Parameters", as shown below:
image

**Comment by robe070** [comment]
When I create a stack from Dev template with commit c917f35fe4a4f958948fd9ee2ff74461a7e0e029 I get this error:
```
Template format error: Unresolved resource dependencies [Git Repository Branch] in the Resources block of the template
```

And the default branch needs to be `support/scalable` which you may need to override to `debug/paas` or include the new source parameter to provider a developer sub directory for unit testing.

**Comment by SakshamAccolite** [comment]
Hi @robe070 , 
Fixed. And changed the branch to support/scalable.

**Comment by robe070** [comment]
Failed checks are understood and unrelated. So merging now.

**Comment by robe070** [comment]
Note: still need to merge in master template changes when they are all complete.

**Pull request merged**